src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -33368,635 +33971,3631 @@ .maxstack  3
+        }
+
+        [Fact]
+        public void AddMethod_Extension_11_WrongThisType()
+        {
+            string source = """
+                using System.Collections;
+                using System.Collections.Generic;
+                class MyCollection<T> : IEnumerable
+                {
+                    private readonly List<T> _list = new();
+                    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
+                    internal void __AddInternal(T t) { _list.Add(t); }
+                }
+                static class Extensions
+                {
+                    public static void Add<T>(this IEnumerable<T> collection, int x) => throw null;
+                }
                 class Program
                 {
                     static void Main()
                     {
-                        object x = new object();
-                        object y = null;
-                        object[] z = new[] { x };
-                        F(z, () => [x]);
-                        F(z, () => [y]);
+                        int x = 1;
+                        int[] y = [2, 3];
+                        MyCollection<int?> z = [x, ..y];
+                        MyCollection<int?> w = new() { x };
                     }
-                    static void F<T>(T t, Func<T> f) { }
                 }
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (8,20): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         object y = null;
-                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(8, 20),
-                // (11,21): warning CS8601: Possible null reference assignment.
-                //         F(z, () => [y]);
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "y").WithLocation(11, 21));
+                    // (19,32): error CS9215: Collection expression type 'MyCollection<int?>' must have an instance or extension method 'Add' that can be called with a single argument.
+                    //         MyCollection<int?> z = [x, ..y];
+                    Diagnostic(ErrorCode.ERR_CollectionExpressionMissingAdd, "[x, ..y]").WithArguments("MyCollection<int?>").WithLocation(19, 32),
+                // (20,40): error CS1061: 'MyCollection<int?>' does not contain a definition for 'Add' and no accessible extension method 'Add' accepting a first argument of type 'MyCollection<int?>' could be found (are you missing a using directive or an assembly reference?)
+                //         MyCollection<int?> w = new() { x };
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "x").WithArguments("MyCollection<int?>", "Add").WithLocation(20, 40)
+                );
         }
 
         [Fact]
-        public void TargetTypedElement_PublicAPI_List()
+        public void AddMethod_Extension_12_WrongThisType()
```

Comment:
Let's have a version of this one with a variant interface: IE, the `Add` method is defined as an extension on `IEnumerable<object>`, when I have a `MyCollection<string>` and it implements `IEnumerable<string>`.

Reply:
Added `AddMethod_Extension_13_WrongThisType`.